scipy/optimize/_sda.py

```diff
@@ -0,0 +1,589 @@
+        lu = list(zip(*bounds))
+        lower = np.array(lu[0])
+        upper = np.array(lu[1])
+        # Checking that bounds are consistent
+        if not np.all(lower < upper):
+            raise ValueError('Bounds are note consistent min < max')
+        # Wrapper for the objective function and minimizer
+        if minimizer_kwargs is None:
+            minimizer_kwargs = dict()
+        self.owf = ObjectiveFunWrapper(bounds, fun, **minimizer_kwargs)
+        # Initialization of RandomState for reproducible runs if seed provided
+        self.rs = check_random_state(seed)
+        # Initialization of the energy state
+        self.es = EnergyState(lower, upper)
+        self.es.reset(self.owf, self.rs, x0)
+        # Maximum number of function call that can be used a stopping criterion
+        self.maxfun = maxfun
+        # Maximum number of step (main iteration)  that can be used as
+        # stopping criterion
+        self.max_steps = max_steps
+        # Minimum value of annealing temperature reached to perform
+        # re-annealing
+        self.temperature_start = temperature_start
+        self.temperature_restart = 0.1
+        # VisitingDistribution instance
+        vd = VisitingDistribution(lower, upper, qv, self.rs)
+        # Markov chain instance
+        self.mc = MarkovChain(qa, vd, self.owf, self.rs, self.es)
+        self.qv = qv
+        self.pure_sa = pure_sa
+
+    def search(self):
+        max_steps_reached = False
+        self._iter = 0
+        t1 = np.exp((self.qv - 1) * np.log(2.0)) - 1.0
+        while(not max_steps_reached):
+            for i in range(self.max_steps):
+                # Compute temperature for this step
+                s = float(i) + 2.0
+                t2 = np.exp((self.qv - 1) * np.log(s)) - 1.0
+                temperature = self.temperature_start * t1 / t2
+                self._iter += 1
+                if self._iter == self.max_steps:
+                    max_steps_reached = True
+                    break
+                # Need a re-annealing process?
+                if temperature < self.temperature_restart:
+                    self.es.reset(self.owf, self.rs)
+                    break
+                # starting Markov chain
+                self.mc.run(i, temperature)
+                if self.owf.nb_fun_call >= self.maxfun:
+                    break
+                if not self.pure_sa:
+                    self.mc.local_search()
+                    if self.owf.nb_fun_call >= self.maxfun:
+                        break
+
+    @property
+    def result(self):
+        """ The OptimizeResult """
+        res = OptimizeResult()
+        res.x = self.es.xbest
+        res.fun = self.es.ebest
+        res.nit = self._iter
+        res.ncall = self.owf.nb_fun_call
+        return res
+
+
+def sda(func, x0, bounds, maxiter=1000, minimizer_kwargs=None,
+        initial_temp=5230., visit=2.62, accept=-5.0, maxfun=1e7, seed=None,
+        pure_sa=False):
+    """
```

Comment:
I think it would be useful to allow the user to have more freedom specifying the local search method parameters. And I think you should also include the gradient of the local search method as an optional argument, and use finite-difference option only when this argument is not provided.

Reply:
I thought that this was I did. Local search method, related arguments and gradient can be passed in "minimizer_kwargs" argument like for basin_hopping, if not specified then "L-BFGS-B" is used.Is this not correct?

Reply:
You are right. I have missed this out. Your implementation does gives this freedom to the user.